### PR TITLE
ovl_Dm_Bal decompiled

### PR DESCRIPTION
**Pull request by sonicdcer**

### CODE REVIEW

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 1]
```diff
@@ -30,31 +36,185 @@ const ActorInit Dm_Bal_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Bal/func_80C1F060.s")
+static AnimationInfo D_80C1F170[0xE] = {
+    { 0x060005FC, 1.0f, 0.0f, 0.0f, 0, -8.0f }, { 0x06000840, 1.5f, 0.0f, 0.0f, 0, -8.0f },
+    { 0x06000840, 1.5f, 0.0f, 0.0f, 2, -4.0f }, { 0x0600A7DC, 1.0f, 0.0f, 0.0f, 2, -4.0f },
+    { 0x0600B1E8, 1.0f, 0.0f, 0.0f, 0, -8.0f }, { 0x0600B604, 1.0f, 0.0f, 0.0f, 0, -8.0f },
+    { 0x0600C498, 1.0f, 0.0f, 0.0f, 2, -8.0f }, { 0x0600C8D8, 1.0f, 0.0f, 0.0f, 0, -8.0f },
+    { 0x0600C8D8, 1.0f, 0.0f, 0.0f, 2, -8.0f }, { 0x0600C498, 1.0f, 23.0f, 0.0f, 2, -8.0f },
+    { 0x0600D530, 1.0f, 0.0f, 0.0f, 0, -8.0f }, { 0x06000C78, 1.0f, 0.0f, 0.0f, 0, -2.0f },
+    { 0x0600CB78, 1.0f, 0.0f, 0.0f, 0, -2.0f }, { 0x06001804, 1.0f, 0.0f, 0.0f, 0, -2.0f },
+};
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:47]: These `0x06XXXXXX` addresses are segment pointers that come from the actors object. Could you change them to use the correct symbols from there. Also the second to last argument (the integer) should use `AnimationModes` enum found in `z64animation.h`. In this case `0 -> ANIMMODE_LOOP` and `2 -> ANIMMODE_ONCE`. Last we usually leave the array length of array declarations, i.e. arrays that have there data declared with them. In this case that would just mean removing the `0xE`.

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 2]
```diff
@@ -30,31 +36,185 @@ const ActorInit Dm_Bal_InitVars = {
...
+        return;
+    }
+    this->unk198 = 0;
+    this->unk334 = 0x3C;
+}
+
+void func_80C1ED64(DmBal* this, GlobalContext* globalCtx, Vec3f* arg2, Vec3f* arg3, f32 arg4) {
+    Actor* paper;
+
+    paper = Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_PAPER, arg2->x, arg2->y, arg2->z, 0, 0, 0, 0);
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:157]:
```suggestion
    Actor* paper = Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_PAPER, arg2->x, arg2->y, arg2->z, 0, 0, 0, 0);
```

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 3]
```diff
@@ -30,31 +36,185 @@ const ActorInit Dm_Bal_InitVars = {
...
+    if (paper != NULL) {
+        paper->velocity = *arg3;
+        paper->gravity = arg4;
+    }
+}
+
+void DmBal_Update(Actor* thisx, GlobalContext* globalCtx) {
+    DmBal* this = THIS;
+    f32 sp44;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:166]:
```suggestion
    s32 pad;
    DmBal* this = THIS;
```
Looks like `sp44` is actually some pad variable.

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 4]
```diff
@@ -30,31 +36,185 @@ const ActorInit Dm_Bal_InitVars = {
...
+        func_80C1ED64(this, globalCtx, &sp3C, &sp30, -0.5f);
+    }
+    this->actionFunc(this, globalCtx);
+    func_80C1EAE8(this, globalCtx);
+    func_80C1EC60(this, globalCtx);
+    func_80C1ED0C(this);
+    SkelAnime_Update(&this->skelAnime);
+}
+
+s32 DmBal_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, Actor* thisx) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:188]: Looks like Elliptic found a better match https://decomp.me/scratch/Tgm0N

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h [hunk 5]
```diff
@@ -8,11 +8,22 @@ struct DmBal;
...
+    /* 0x144 */ SkelAnime skelAnime;
+    /* 0x188 */ DmBalActionFunc actionFunc;
+    /* 0x18C */ f32 unk18C;
+    /* 0x190 */ f32 unk190;
+    /* 0x194 */ f32 unk194;
+    /* 0x198 */ s32 unk198;
+    /* 0x19C */ Vec3s morphTable;
+    /* 0x1A2 */ char pad1A2[0xC0];
+    /* 0x262 */ Vec3s joinTable;
+    /* 0x268 */ char pad268[0xCC];
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h:21]: Same comment as in the `Hata` pr, https://github.com/zeldaret/mm/pull/844#discussion_r891677256

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c (hunk 1 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:47]: done

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h [hunk 6]
```diff
@@ -2,17 +2,28 @@
...
-} DmBal; // size = 0x33C
+    /* 0x000 */ Actor actor;
+    /* 0x144 */ SkelAnime skelAnime;
+    /* 0x188 */ DmBalActionFunc actionFunc;
+    /* 0x18C */ f32 unk_18C;
+    /* 0x190 */ f32 unk_190; 
+    /* 0x194 */ f32 unk_194;
+    /* 0x198 */ s32 unk_198;
+    /* 0x19C */ Vec3s morphTable[OBJECT_BAL_LIMB_MAX];
+    /* 0x262 */ Vec3s joinTable[OBJECT_BAL_LIMB_MAX];
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h:20]:
```suggestion
    /* 0x19C */ Vec3s jointTable[OBJECT_BAL_LIMB_MAX];
    /* 0x262 */ Vec3s morphTable[OBJECT_BAL_LIMB_MAX];
```
Looks like the `jointTable` and `morphTable` are reversed. Also `jointTable` looks to be missing a `t`.

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h [hunk 7]
```diff
@@ -2,17 +2,28 @@
...
+    /* 0x328 */ char pad_328[12];
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.h:21]:
```suggestion
    /* 0x328 */ UNK_TYPE1 pad_328[12];
```

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 8]
```diff
@@ -30,31 +36,178 @@ const ActorInit Dm_Bal_InitVars = {
...
+        this->actor.home.pos = this->actor.world.pos;
+    } else {
+        this->unk_336 = 0;
+        this->unk_198 = 0;
+        D_80C1F2C0 = 0x63;
+    }
+}
+
+Vec3f D_80C1F2C4 = { 0.0f, 9.0f, 0.0f };
+u32 D_80C1F2D0[] = { &object_bal_Tex_006050, &object_bal_Tex_0094D0 };
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:123]:
```suggestion
TexturePtr D_80C1F2D0[] = { object_bal_Tex_006050, object_bal_Tex_0094D0 };
```
Now that we see that these are textures, we should use the `TexturePtr` type as well removing the `&`s.

**Review by hensldm — COMMENTED** [review]
A few more things.

**Review by hensldm — APPROVED** [review]
LGTM, just one more minor thing.

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 9]
```diff
@@ -30,31 +36,178 @@ const ActorInit Dm_Bal_InitVars = {
...
+
+void DmBal_PostLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, Actor* thisx) {
+}
+
+void DmBal_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    DmBal* this = THIS;
+
+    OPEN_DISPS(globalCtx->state.gfxCtx);
+    func_8012C28C(globalCtx->state.gfxCtx);
+    gSPSegment(POLY_OPA_DISP++, 0x08, SEGMENTED_TO_VIRTUAL(D_80C1F2D0[this->unk_198]));
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:209]: Turns out from looking at the textures that these are eye textures, so you can name the array `sEyeTextures` and `unk_198` `eyeIndex`.

**Review by EllipticEllipsis — COMMENTED** [review]
Good job on this, there's some easy documentation we can get done here though.

(Wow, a whole actor for the Branch And Link instruction!)

File: src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c [hunk 10]
```diff
@@ -159,6 +147,8 @@ void func_80C1ED64(DmBal* this, GlobalContext* globalCtx, Vec3f* arg2, Vec3f* ar
     }
 }
 
+Vec3f D_80C1F2C4 = { 0.0f, 9.0f, 0.0f };
+
 void DmBal_Update(Actor* thisx, GlobalContext* globalCtx) {
     DmBal* this = THIS;
     s32 pad;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Dm_Bal/z_dm_bal.c:154]:
```suggestion
    s32 pad;
    DmBal* this = THIS;
```
if it matches.

**Review by EllipticEllipsis — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**